go/libraries/doltcore/schema/index_coll.go

```diff
@@ -0,0 +1,241 @@
+	}
+	ixc.indexes[indexName] = index
+	for _, tag := range tags {
+		ixc.colTagToIndex[tag] = append(ixc.colTagToIndex[tag], index)
+	}
+	return index, nil
+}
+
+func (ixc *indexCollectionImpl) AllIndexes() []Index {
+	indexes := make([]Index, len(ixc.indexes))
+	i := 0
+	for _, index := range ixc.indexes {
+		indexes[i] = index
+		i++
+	}
+	return indexes
+}
+
+func (ixc *indexCollectionImpl) BestMatch(cols ...string) Index {
+	//TODO: actually determine the best index based on the given columns, which can include non-indexed columns
+	if len(cols) == 0 {
+		return nil
+	}
+	match := ixc.colTagToIndex[ixc.colToTag[cols[0]]]
+	if len(match) != 0 {
+		return match[0]
+	}
+	return nil
+}
+
+func (ixc *indexCollectionImpl) Contains(indexName string) bool {
+	_, ok := ixc.indexes[indexName]
+	return ok
+}
+
+func (ixc *indexCollectionImpl) ContainsColumnCollection(cols ...string) bool {
+	tags := make([]uint64, len(cols))
+	for i, col := range cols {
+		tag, ok := ixc.colToTag[col]
+		if !ok {
+			return false
+		}
+		tags[i] = tag
+	}
+	return ixc.ContainsColumnTagCollection(tags...)
+}
+
+func (ixc *indexCollectionImpl) ContainsColumnTagCollection(tags ...uint64) bool {
+	tagCount := len(tags)
+	for _, idx := range ixc.indexes {
+		if tagCount == len(idx.tags) {
+			allMatch := true
+			for i, idxTag := range idx.tags {
+				if tags[i] != idxTag {
+					allMatch = false
+					break
+				}
+			}
+			if allMatch {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (ixc *indexCollectionImpl) Count() int {
+	return len(ixc.indexes)
+}
+
+func (ixc *indexCollectionImpl) Get(indexName string) Index {
+	return ixc.indexes[indexName]
+}
+
+func (ixc *indexCollectionImpl) DropIndex(indexName string) (Index, error) {
+	if !ixc.Contains(indexName) {
+		return nil, fmt.Errorf("`%s` does not exist as an index for this table", indexName)
+	}
+	index := ixc.indexes[indexName]
+	delete(ixc.indexes, indexName)
+	for _, tag := range index.tags {
+		indexesRefThisCol := ixc.colTagToIndex[tag]
+		for i, comparisonIndex := range indexesRefThisCol {
+			if comparisonIndex == index {
+				indexesRefThisCol = append(indexesRefThisCol[:i], indexesRefThisCol[i+1:]...)
+				break
+			}
+		}
+	}
+	return index, nil
+}
+
+func (ixc *indexCollectionImpl) HasIndexes() bool {
+	if len(ixc.indexes) > 0 {
+		return true
+	}
+	return false
+}
+
+func (ixc *indexCollectionImpl) RenameIndex(oldName, newName string) (Index, error) {
+	if !ixc.Contains(oldName) {
+		return nil, fmt.Errorf("`%s` does not exist as an index for this table", oldName)
+	}
+	if ixc.Contains(newName) {
+		return nil, fmt.Errorf("`%s` already exists as an index for this table", newName)
+	}
+	index := ixc.indexes[oldName]
+	delete(ixc.indexes, oldName)
+	index.name = newName
+	ixc.indexes[newName] = index
+	return index, nil
+}
+
+func combineAllTags(tags []uint64, pks []uint64) []uint64 {
+	allTags :=  make([]uint64, len(tags))
+	_ = copy(allTags, tags)
+	foundPKCols := make(map[uint64]bool)
```

Comment:
This is kind of convoluted. You could just put all tags in a boolean map with value true, then only add the pks that aren't in the map. The middle loop is unnecessary.